Perform variable expansion using text.Transformer

drafnel opened the pull request:
Currently NewYAMLProviderFromFiles() and NewYAMLProviderWithExpand() do not produce the same result when given the same input, even when the input contains only simple values with no expansion variables.  The expansion feature has a side-effect which converts all fields into strings.

This can be demonstrated with:


    package main

    import "fmt"
    import "os"

    import "go.uber.org/config"

    func main() {
        if len(os.Args) != 3 {
            fmt.Fprintf(os.Stderr, "Usage: %s config.yaml field\n",
                os.Args[0])
            os.Exit(1)
        }

        filename := os.Args[1]
        field := os.Args[2]

        cfg, _ := config.NewYAMLProviderFromFiles(filename)
        val := cfg.Get(field)
        fmt.Printf("%s: %v (%T)\n", field, val.Value(), val.Value())

        cfg, _ = config.NewYAMLProviderWithExpand(os.LookupEnv, filename)
        val = cfg.Get(field)
        fmt.Printf("%s: %v (%T)\n", field, val.Value(), val.Value())
    }

    $ cat test.yaml
    foo:
        bar:
            baz: 13`

    $ ./get-config test.yaml foo.bar.baz
    foo.bar.baz: 13 (int)
    foo.bar.baz: 13 (string)

This series introduces a text.Transformer to perform the variable expansion via an io.ReadCloser so it can be transparently used by newYAMLProviderCore().  It also introduces the escape sequence '$$' which will be replaced by a literal '$'.

fyi @akshayjshah @breerly 

-Brandon

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber-go/config?pullRequest=58) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber-go/config?pullRequest=58) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber-go/config?pullRequest=58) it.</sub>

alsamylkin commented:
Thank you for the PR Brandon! I really like your fix on expanding strings started with $.(for casting we recommend to use [spf13/cast library](https://github.com/uber-go/config/blob/master/example_cast_test.go#L31)) I'll do a more thorough review tomorrow morning, but there are couple small nits I see now:
1. Please update the changelog.
2. Could you add more tests for transform: strings that ends with `$`, `$$`, consist only of `$$$$$$$`, empty strings?

alsamylkin commented on expand.go:
```diff
@@ -0,0 +1,206 @@
+	rc      io.ReadCloser
+}
+
+func newExpandReadCloser(rc io.ReadCloser, expand_func func(string) (string, error)) *expandReadCloser {
```
please, use camelCase for arguments.

drafnel replied:
Will do.

alsamylkin commented on yaml_test.go:
```diff
@@ -1745,6 +1748,31 @@ func TestNewYamlProviderWithExpand(t *testing.T) {
+	assert.IsType(t, "string", p.Get("a-string").Value())
+	assert.Exactly(t, p.Get("a-string").Value(), pp.Get("a-string").Value())
+}
+
```
Could you add more tests for expand:
```golang
p, err := NewStaticProviderWithExpand(
	map[string]interface{}{
		"nil":   "a",
		"one":   "$a",
		"two":   "$$a",
		"three": "$$$a",
		"four":  "$$$$a",
		"five":  "$$$$$a",
	},
	func(key string) (string, bool) {
		return fmt.Sprint(len(key)), true
	})

require.NoError(t, err)
assert.Equal(t, "a", p.Get("nil").String())
assert.Equal(t, "1", p.Get("one").String())
assert.Equal(t, "$a", p.Get("two").String())
assert.Equal(t, "$1", p.Get("three").String())
assert.Equal(t, "$$a", p.Get("four").String())
assert.Equal(t, "$$1", p.Get("five").String())
```
 I am pretty sure, it will pass, but I want to keep them for future refactoring(if any) to be easy.

drafnel replied:
Thanks for these.

drafnel commented:
So how do I update my commits in github?  Can I just force push to the branch in my repo?

alsamylkin commented:
Yes, I think a regular push to your branch will be enough.

drafnel commented:
Cool, that worked.  Series updated.

breerly commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v1.0.2 (unreleased)
```
Looks like we're introducing new functionality here? This should be `v1.1.0` in that case.

drafnel replied:
Agreed.

I consider most of this to be either a tightening of the existing rules (e.g. rules for variable names) or a correction of less-than-desired behavior with respect to the underlying data types of Value objects that are produced when variable substitution is active.

The only part I'd really consider a new feature is the '$$' handling.

Any of that is enough to justify more than a patch-level version bump.

drafnel commented:
Noticed a bug after rereading the text.transform docs.  nil error should only be returned from the Transform function if all of the bytes in src were consumed.  Series updated to 300123e.  Here's what the interdiff looks like:

    $ git diff 32aa395 300123e
    diff --git a/expand.go b/expand.go
    index da068e0..4524827 100644
    --- a/expand.go
    +++ b/expand.go
    @@ -87,7 +87,12 @@ func (this *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int,
            var i int
            var j int

    -       for i < len(src) && j < len(dst) {
    +       for i < len(src) {
    +
    +               if j == len(dst) {
    +                       return j, i, transform.ErrShortDst
    +               }
    +
                    // Not a token-starting character, just copy it
                    if src[i] != '$' {
                            dst[j] = src[i]

drafnel commented:
Just a few last tweaks to prefer bytes.IndexByte() and copy() where possible, and simplify the way bare variable names are parsed.  I don't expect to make any more changes.  Here's the interdiff against the last update:

    diff --git a/expand.go b/expand.go
    index 4524827..2c03721 100644
    --- a/expand.go
    +++ b/expand.go
    @@ -70,10 +70,11 @@ func isShellNameChar(c byte) bool {
     		(c >= '0' && c <= '9')
     }
     
    -// bytesIndexFunc is what bytes.IndexFunc should be
    -func bytesIndexFunc(buf []byte, f func(b byte) bool) int {
    +// bytesIndexCFunc returns the index of the byte for which the
    +// complement of the supplied function is true
    +func bytesIndexCFunc(buf []byte, f func(b byte) bool) int {
     	for i := 0; i < len(buf); i++ {
    -		if f(buf[i]) {
    +		if !f(buf[i]) {
     			return i
     		}
     	}
    @@ -93,14 +94,29 @@ func (this *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int,
     			return j, i, transform.ErrShortDst
     		}
     
    -		// Not a token-starting character, just copy it
    -		if src[i] != '$' {
    -			dst[j] = src[i]
    -			i++
    -			j++
    +		end := bytes.IndexByte(src[i:], '$')
    +
    +		if end == -1 {
    +			// src does not contain '$', copy into dst
    +			cnt := copy(dst[j:], src[i:])
    +			i += cnt
    +			j += cnt
     			continue
    +		} else if end > 0 {
    +			// copy chars preceding '$' from src to dst
    +			cnt := copy(dst[j:], src[i:i+end])
    +			i += cnt
    +			j += cnt
    +
    +			if j == len(dst) {
    +				return j, i, transform.ErrShortDst
    +			}
    +
    +			// fallthrough
     		}
     
    +		// src[i] now points to '$', j < len(dst)
    +
     		// If we're at the end of src, but we found a starting
     		// token, return ErrShortSrc, unless we're also at EOF,
     		// in which case just copy it dst.
    @@ -126,7 +142,6 @@ func (this *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int,
     		}
     
     		var token []byte
    -		var end int
     
     		// Start of bracketed token ${foo}
     		if src[i+1] == '{' {
    @@ -168,9 +183,7 @@ func (this *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int,
     				continue
     			}
     
    -			end = bytesIndexFunc(src[i+2:], func(c byte) bool {
    -				return !isShellNameChar(c)
    -			})
    +			end = bytesIndexCFunc(src[i+2:], isShellNameChar)
     
     			if end == -1 {
     				// Reached the end of src without finding

drafnel commented:
Hey @alsamylkin is there anything holding this up?

alsamylkin commented on expand_test.go:
```diff
@@ -0,0 +1,244 @@
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	for _, tst := range tests {
```
Could you use [subtests](https://blog.golang.org/subtests) instead of table driven tests?

drafnel replied:
Sure.

alsamylkin commented on expand.go:
```diff
@@ -0,0 +1,224 @@
+
+import (
+	"bytes"
+	"golang.org/x/text/transform"
```
Looks like this import should be in a separate group.

drafnel replied:
@glibsm no I don't use `go imports`.  I prefer not to have tools rewriting my code behind my back.

akshayjshah replied:
That's absolutely your prerogative, but you'll get less grief if you take a little time to internalize the formatting that it imposes. Many projects will require that your code conform to the same style.

drafnel commented on expand.go:
```diff
@@ -0,0 +1,224 @@
+
+import (
+	"bytes"
+	"golang.org/x/text/transform"
```
I couldn't care less.  So you're saying you want a blank line between these two imports?

glibsm replied:
```
import (
  // stdlib

 // current project

 // external
)
```

Do you have `go imports` configured for your editor?

alsamylkin replied:
@drafnel  yes just add a blank line. Somehow `expand_test.go` picked it up correctly, but `expand.go` didn't.

drafnel commented on expand.go:
```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+func (this *expandReadCloser) Read(p []byte) (int, error) {
```
I consider that advice from the go project to be the worst kind of advice.  "Use single letter variable names".  Garbage.  When you see the letter 'b' in a function, should you immediately assume that it refers to the calling object?  What about the letter 'c'?  I often use the letter 'c' for a variable that holds a single character.  However, if you see a variable named "this", every developer will immediately understand that it refers to the calling object.

Now, I understand that "methods" in Go are much more similar to the way you do objects in C, by passing in the object as the first argument.  So while a Go method definition might look like this:

   func (*ObjectType) foo (...)

it is equivalent to a function defined like this:

   func foo (*ObjectType, ...) 

So what?  It does not improve readability or remove ambiguity by using a variable named 'o' to refer to the object which called this function.  Why Rob Pike thinks this is a good idea, and actively rails against using something that I consider both more readable and less ambiguous, is beyond me.

Python has the same calling convention for methods as Go does, i.e. the calling object is passed in as the first argument and can be given any name by the developer.  At least that project had the sense to understand that there is zero value in using a different name for the calling object in every different class, and there is a negative impact on readability, so they suggest calling it "self".  That's a convention worth documenting IMHO.

But, it's your project so I'll use whatever letter you want me to use.  I can't think of a single letter that would be immediately recognized by a random developer as referring to the calling object.  Please tell me which letter you'd like me to use.

alsamylkin replied:
I don't really care about the rule, some linters are using all the rules from the wiki to validate code(e.g. my VSCode) so I just want to keep them happy. Lets use `e` for the `expandReadCloser` struct.

drafnel commented on expand_test.go:
```diff
@@ -0,0 +1,244 @@
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	for _, tst := range tests {
```
Sigh...

drafnel commented on expand.go:
```diff
@@ -0,0 +1,224 @@
+
+// expandReadCloser implements io.ReadCloser
+type expandReadCloser struct {
+	tReader *transform.Reader
```
This is exactly what I wanted to do in the first place, but I didn't know I could.  Thanks!

drafnel commented on CHANGELOG.md:
```diff
@@ -1,5 +1,24 @@
 
+## v1.1.0 (unreleased)
+
+- Expand variables (e.g. $foo ${foo}) using a text.Transform instead
```
Sure, I can add this detail to the NewYAMLProviderWithExpand docs.

drafnel commented on expand_test.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func TestExpander(t *testing.T) {
+	var buf []byte
```
Sure.

alsamylkin commented:
@breerly, @akshayjshah are you ok with these changes?

drafnel commented:
> Most importantly, please sign the CLA!

Didn't I already sign this when I submitted to thriftrw?  Do I need to do this for every single repo?

> We can fix the unintended type changes by adding another case to the type switch in recursiveApply and only expanding strings. This has the added benefit of not making the final type of n: ${FOO} depend on the value of $FOO - it's always a string, which matches the current behavior (and is IMO more reasonable).

I don't find that behavior more reasonable.  Why should it be reasonable for a value of '13' to sometimes be a string and sometimes be an integer?

I would find it less surprising (and hence more reasonable) if '13' always produced an integer regardless of whether the field in the yaml file was specified like:

    foo: 13

or like:

    foo: ${FOO:13}

Once you have defined behavior that is also consistent, it allows the call-sites to be simpler.

> We can support the $-escaping, using exactly the rules you've proposed, by special-casing in == "$" in the anonymous function in replace. See https://play.golang.org/p/Rl2ywc3yfv for an example of this working.

Yup, that's true.  It's not necessarily the right place to put it depending on how you draw your boundaries of concern though.

What you don't get with either of your suggestions is an elegant solution in my opinion.  io.Reader is a natural interface for this and allows the code within yaml.go to be nicely simplified since the io.Reader provided by expand.go can be wrapped around the file readers and then passed to the lower level functions within yaml.go.  You have a net reduction in code paths since you don't have to have special cases for the "WithExpand" case.

Ideally, the yaml parser would accept an io.Reader too instead of requiring the caller to do an io.ReadAll() and then pass in the byte slice.  At some point I expect such a yaml parser will be created and yaml.go can be simplified further.

-Brandon

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// First char of shell variable may be [a-zA-Z_]
+func isShellNameFirstChar(c byte) bool {
```
Looking at the implementation of os.Expand, it looks like it supports `$1` and `$1foo`. Should we replicate that behavior?

https://play.golang.org/p/3g_9wwuQU_

drafnel replied:
Well, `$1foo` would be parsed as `$1` followed by the literal characters `foo`.  In shell context $1, $2, etc refer to parameter arguments,  $* refers to the concatenated parameter arguments, $@ refers to the quoted parameter arguments, $$ refers to the process id, etc.  They all have specific meaning in a shell context which don't really seem to make sense here.  I'd stick strictly to shell variable names which have the form "[a-zA-Z_][a-zA-Z0-9_]*".  But I don't feel too strongly about it.

Of course, the code I've uploaded doesn't strictly parse variables wrapped in curly braces, so it's possible that someone could use ${1} as a variable name.  Not that that's a good thing, just pointing it out.

abhinav replied:
I don't feel strongly about whether we should support `$1` or not. It won't be a valid environment variable for sure. I'm happy to leave that as-is.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+func bytesIndexCFunc(buf []byte, f func(b byte) bool) int {
+	for i := 0; i < len(buf); i++ {
+		if !f(buf[i]) {
+			return i
```
optional nit: why use this instead of `range`?

```
for i, b := range buf {
  if !f(b) {
    return i
  }
}
```

drafnel replied:
No reason, just didn't occur to me.  I don't feel strongly about either form, so I can use range if you prefer.

abhinav replied:
Yeah, please change it to range. Thanks.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+// a literal '$'.
+func (e *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int, error) {
+	var i int
+	var j int
```
nit: can we use the names used in the interface? https://godoc.org/golang.org/x/text/transform#Transformer

The names nDst and nSrc are more self-documenting than i and j.

drafnel replied:
Sure.  I have a mild resistance to the 'n' prefix, which is generally reserved for holding a "count" or "number" of X, but I'll change it.

abhinav replied:
I'm not committed to the naming as done by the Transformer interface. Similar variable names that are more descriptive would work too (srcPos, dstPos).

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+				return j, i, transform.ErrShortDst
+			}
+
+			// fallthrough
```
comment not necessary. fallthrough is expected.

drafnel replied:
Yes, the comment is purely to communicate to the reader that the fallthrough is intentional in the 'else' block, unlike the 'if' portion of the block which calls continue.  Happy to delete it if it does not improve readability.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+
+		end := bytes.IndexByte(src[i:], '$')
+
+		if end == -1 {
```
Nit: inline the declaration of `end` into the if-else block. It's not used
outside it.

```
if end := bytes.IndexByte(...); end == -1 {
  // ..
} else if end > 0 {
  // ..
}
```

The intention is to keep the scope of the variable as small as possible.

drafnel replied:
Sure.

drafnel replied:
Hmm, I think I find

    if end := bytes.IndexByte(src[srcPos:], '$'); end == -1 {
        ...
    } else if end > 0 {

to be considerably less readable than

    end := bytes.IndexByte(src[srcPos:], '$')
    if end == -1 {
        ...
    } else if end > 0 {

Maybe the notation will grow on me.  For now, I'd prefer to leave it as is.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+	for i < len(src) {
+
+		if j == len(dst) {
+			return j, i, transform.ErrShortDst
```
This and the other "ErrShortDst/ErrShortSrc" cases don't have coverage. These look like important corner cases we should have tests for.

drafnel replied:
You may have been looking at an older version of this commit.  The current commit should have 100% test coverage of expand.go.  A couple of these were particularly difficult to trigger and require intimate knowledge of the text/transform package.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+		// Start of bracketed token ${foo}
+		if src[i+1] == '{' {
+			end = bytes.IndexByte(src[i+2:], '}')
+			if end == -1 {
```
Once the other `end` is out of scope, you should declare this one right above
this `if` block with a slightly more descriptive name.  Something like,
`varEnd` or `tokenEnd` maybe, or a comment to go with it.

```
var tokenEnd int  // end of the $foo or ${foo}
if ... {
  // ...
  tokenEnd = ...
} else {
  // ...
  tokenEnd = ...
}
```

It makes it much clearer that `tokenEnd` is the output of the following if-
else block.

drafnel replied:
Makes sense.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+					// We need more bytes in src
+					return j, i, transform.ErrShortSrc
+				}
+				end = len(src)
```
Can we add a test where `$foo` is at the end of the string (so at EOF). That'll cover this line.

drafnel replied:
Already exists.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+			return j, i, err
+		}
+
+		cnt := copy(dst[j:], replacement)
```
Shouldn't you make sure there's enough room in `dst` before copying into it? If the replacement is too large, it'll be copied partially, then copied again when this is retried after `ErrShortDst` is returned.

drafnel replied:
The copy itself was being used to check whether there was enough room.  I expect the normal case to be that there is enough room, but sure, we can the length and then avoid the unnecessary copy here and above.

abhinav replied:
copy will copy at most `len(dst)` bytes, right? The issue I was pointing out was that in the case where `dst` is too short, copy will copy some number of bytes less than `len(src)`, we'll return `ErrDstShort`, in which case text/transform will retry after resizing dst, and we'll copy the part of `src` that was already copied again. (This all, assuming I understand the contract of text/transform correctly.)

drafnel replied:
Yes that's correct.  I think I understood your original comment.  In my reply when I referred to "the unnecessary copy", I was talking about the case you described, the first copy that would happen when dst was too short, and which would then be performed again in full after dst was resized.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,217 @@
+		// If we couldn't copy the replacement string into dst,
+		// we need more room in dst.
+
+		return j, i, transform.ErrShortDst
```
Can we test for this by providing a buffer that is too short for a very large replacement string?

drafnel replied:
There should already be a test that triggers this code path.  Ultimately the text/transform package currently uses a buffer that is 4096 bytes, so a substitution may not be longer than that.

drafnel commented:
Ok, I think I addressed all of the comments.  Let me know if I missed anything.  I'll hold off on resolving the merge until we decide how to handle pull request #64 which I'd like to slip in after the first commit in this series which would allow me to delete the expandReadCloser object from expand.go.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,213 @@
+// a literal '$'.
+func (e *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int, error) {
+	var srcPos int
+	var dstPos int
```
nit: you can use `var srcPos, dstPos int` here. The "type after names" syntax is there for exactly this reason.

drafnel replied:
I don't think that improves readability, so, perhaps with some exceptions, I tend to use one declaration per line.

abhinav commented on expand.go:
```diff
@@ -0,0 +1,213 @@
+		}
+
+		var token []byte
+		var tokenEnd int
```
nit: var groups are cool

```go
var (
  token []byte
  tokenEnd int
)
```

drafnel replied:
Oh let me count the multitude of ways by which a variable can be defined.  I think Rob Pike forgot that your design isn't done when there's nothing left to add, it's done when there's nothing left to take away :-)

drafnel commented:
Ok, I inserted the commit from #64 into this series like I mentioned before.  If everyone's happy, I think this is ready to merge.

alsamylkin commented:
Please don't insert it, until the #64 is approved.
Both changes will be released together, no need to merge PR.

abhinav commented:
Regarding moving the changes into this PR:

What @drafnel intends for us to do here is to merge the PR without squashing.
If you look at the individual commits, they're all highly detailed commit
messages. If we merge the changes as-is without squashing, the PRs will be
closed as expected and the commits will tell a story similar to if @drafnel
created 5 different PRs.

I think the two PRs should be fine to merge together once #64 gets approved.

abhinav commented:
I'm merging these changes in. Writing tests for this specific case is coverage
for the sake of coverage and would require more refactoring than it's worth.

abhinav commented:
I'm merging these changes in. Writing tests for this specific case is coverage
for the sake of coverage and would require more refactoring than it's worth.
